pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java

```diff
@@ -26,15 +26,38 @@
 public class HttpSegmentFetcher implements SegmentFetcher {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(HttpSegmentFetcher.class);
+  private static final String MAX_RETRIES = "maxRetries";
+  private static final int DEFAULT_MAX_RETRIES = 3;
+  private int maxRetries = DEFAULT_MAX_RETRIES;
 
   @Override
   public void init(Map<String, String> configs) {
+    if (configs.containsKey(MAX_RETRIES)) {
+      try {
+        maxRetries = Integer.parseInt(configs.get(MAX_RETRIES));
+      } catch (Exception e) {
+        maxRetries = DEFAULT_MAX_RETRIES;
+      }
+    }
   }
 
   @Override
   public void fetchSegmentToLocal(String uri, File tempFile) throws Exception {
-    final long httpGetResponseContentLength = FileUploadUtils.getFile(uri, tempFile);
-    LOGGER.info("Downloaded file from {} to {}; Length of httpGetResponseContent: {}; Length of downloaded file: {}", uri, tempFile,
-        httpGetResponseContentLength, tempFile.length());
+    for (int retry = 1; retry <= maxRetries; ++retry) {
```

Comment:
The segment refresh message IS asynchronous. If we handle multiple refresh messages at the same time, we can run out of memory on the servers, since a refreshed segment is not unloaded until all the queries are completed for that segment.

Further, the entire helix pipeline is not blocked on a refresh. Only other segment refreshes are blocked. The rest of the helix transition, etc. will happen as usual.

Ideally, we need to differentiate between a download failure vs load failure (exception on load). The former is intermittent, and the latter is probably permanent, and there is no reason to retry.

In that sense, adding a retry for HttpFetcher (for the server part) seems fine. We need to change the logic that calls this class and handle  things appropriately.